ompi/mca/coll/base/coll_base_alltoallv.c

```diff
@@ -217,32 +227,72 @@ ompi_coll_base_alltoallv_intra_pairwise(const void *sbuf, const int *scounts, co
     ompi_datatype_type_size(sdtype, &sdtype_size);
     ompi_datatype_type_size(rdtype, &rdtype_size);
 
-    if (0 == sdtype_size || 0 == rdtype_size) {
+    if ((0 == sdtype_size || 0 == rdtype_size) && !SEND_0BYTES_ANYWAYS) {
         /* Nothing to exchange */
         return MPI_SUCCESS;
     }
 
     ompi_datatype_type_extent(sdtype, &sext);
     ompi_datatype_type_extent(rdtype, &rext);
 
-   /* Perform pairwise exchange starting from 1 since local exchange is done */
-    for (step = 0; step < size; step++) {
+    /* Simple optimization - handle send to self first */
+    psnd_left = ((char *) sbuf) + (ptrdiff_t)sdisps[rank] * sext;
+    prcv_left = ((char *) rbuf) + (ptrdiff_t)rdisps[rank] * rext;
+    if (0 < scounts[rank]) {
+        err = ompi_datatype_sndrcv(psnd_left, scounts[rank], sdtype,
+                                   prcv_left, rcounts[rank], rdtype);
+        if (MPI_SUCCESS != err) {line = __LINE__; goto err_hndl;}
+    }
+
+      /* Perform pairwise exchange starting from 1 since local exchange is done */
+    for (step = 1; step <= (size >> 1); step++) {
 
-        /* Determine sender and receiver for this step. */
-        sendto  = (rank + step) % size;
-        recvfrom = (rank + size - step) % size;
+        /* this sequence must be compatible with the inplace version,
+           since do not know if other ranks are in-place or not. */
+        left = (rank + step) % size; /* (sendrecv) */
+        right  = (rank + size - step) % size; /* (recv, then send) */
 
         /* Determine sending and receiving locations */
-        psnd = (char*)sbuf + (ptrdiff_t)sdisps[sendto] * sext;
-        prcv = (char*)rbuf + (ptrdiff_t)rdisps[recvfrom] * rext;
-
-        /* send and receive */
-        err = ompi_coll_base_sendrecv( psnd, scounts[sendto], sdtype, sendto,
-                                        MCA_COLL_BASE_TAG_ALLTOALLV,
-                                        prcv, rcounts[recvfrom], rdtype, recvfrom,
-                                        MCA_COLL_BASE_TAG_ALLTOALLV,
-                                        comm, MPI_STATUS_IGNORE, rank);
-        if (MPI_SUCCESS != err) { line = __LINE__; goto err_hndl;  }
+        psnd_left = (char*)sbuf + (ptrdiff_t)sdisps[left] * sext;
+        prcv_left = (char*)rbuf + (ptrdiff_t)rdisps[left] * rext;
+        psnd_right = (char*)sbuf + (ptrdiff_t)sdisps[right] * sext;
+        prcv_right = (char*)rbuf + (ptrdiff_t)rdisps[right] * rext;
+
+        if (0 != rcounts[right]*rdtype_size || SEND_0BYTES_ANYWAYS) {
```

Comment:
`count * dtype_size` is not very efficient. `0 < count && 0 < dtype_size` should be better.